packages/fast-usdc/src/exos/advancer.js

```diff
@@ -9,116 +11,219 @@ import { addressTools } from '../utils/address.js';
+    'Fast USDC Advancer',
     {
-      /**
-       * @param {undefined} result TODO confirm this is not a bigint (sequence)
-       * @param {{ destination: ChainAddress; amount: bigint; }} ctx
-       */
-      onFulfilled(result, { destination, amount }) {
-        log(
-          'Advance transfer fulfilled',
-          q({ amount, destination, result }).toString(),
-        );
-      },
-      onRejected(error) {
-        // XXX retry logic?
-        // What do we do if we fail, should we keep a Status?
-        log('Advance transfer rejected', q(error).toString());
-      },
+      advancer: M.interface('AdvancerI', {
+        handleTransactionEvent: M.callWhen(CctpTxEvidenceShape).returns(
+          M.or(M.undefined(), VowShape),
+        ),
+      }),
+      ...WatcherHandlersShape,
     },
-  );
-
-  return zone.exoClass(
-    'Fast USDC Advancer',
-    M.interface('AdvancerI', {
-      handleTransactionEvent: M.call(CctpTxEvidenceShape).returns(VowShape),
-    }),
     /**
      * @param {{
+     *   assetManagerFacet: AssetManagerFacet;
      *   localDenom: Denom;
-     *   poolAccount: HostInterface<OrchestrationAccount<{ chainId: 'agoric' }>>;
+     *   poolAccount: ERef<HostInterface<OrchestrationAccount<{chainId: 'agoric';}>>>
+     *   usdcBrand: Brand<'nat'>;
      * }} config
      */
     config => harden(config),
     {
-      /** @param {CctpTxEvidence} evidence */
-      handleTransactionEvent(evidence) {
-        // TODO EventFeed will perform input validation checks.
-        const { recipientAddress } = evidence.aux;
-        const { EUD } = addressTools.getQueryParams(recipientAddress).params;
-        if (!EUD) {
-          statusManager.observe(evidence);
-          throw makeError(
-            `recipientAddress does not contain EUD param: ${q(recipientAddress)}`,
-          );
-        }
+      advancer: {
+        /**
+         * Returns a Promise for a Vow in the happy path, or undefined
+         * when conditions are not met. Aims to perform a status update for
+         * every observed transaction.
+         *
+         * We do not expect any callers to depend on the settlement of
+         * `handleTransactionEvent` - errors caught are communicated to the
+         * `StatusManager` - so we don't need to concern ourselves with
+         * preserving the vow chain for callers.
+         *
+         * @param {CctpTxEvidence} evidence
+         */
+        async handleTransactionEvent(evidence) {
+          await null;
+          try {
+            // TODO poolAccount might be a vow we need to unwrap
+            const { assetManagerFacet, poolAccount, usdcBrand } = this.state;
+            const { recipientAddress } = evidence.aux;
+            const { EUD } =
+              addressTools.getQueryParams(recipientAddress).params;
+            if (!EUD) {
+              throw makeError(
+                `recipientAddress does not contain EUD param: ${q(recipientAddress)}`,
+              );
+            }
```

Comment:
We must also fail if there are unexpected values, which could have been used to mislead the user. E.g. `Destination=ThePlaceYouWant&EUD=AttackerWallet`

Define a shape for the expected params. You could use use a `mustMatch` to get the type narrowing, or maybe even better is make `getQueryParams` take a shape for the params.

Incidentally, I just noticed that `getQueryParams` returns the address too, which is confusing. I also notice that no calls yet use that key. So I propose that `getQueryParams(addr, shape)` returns the params object directly, having asserted it matches the shape. Then the caller can use the object as typed.

I'm happy to make that change in a new PR if you like

Reply:
Ok. Can we leave this level of verification for the `Feed`? This is where I thought we were planning to do evidence validation.

Reply:
they both can. The reason to have the shape for `getQueryParams` is because the type passed in just has `address` as a string. This code needs some what to assert to the type system that `EUD` came out.

Reply:
Ok I think I see. PTAL at the last commit on the PR.